Pagination typehead

nicolearagao opened the pull request:
- Add TypeaheadCheckboxesWithSearch component with server-side filtering
- Implement debounced search (300ms) to prevent excessive API requests
- Add pagination support with hasMorePages detection
- Maintain selected credential state across search operations
- Add useGetCredentialsWithSearchApi hook for server-side search API calls
- Replace local filtering with server-side search_by_name parameter
- Support credential cache to handle selected items across pages
- Update Add Source modal to use new search-enabled typeahead component

Resolves issue where UI could only filter locally-known credentials,
making it impossible to select from large credential sets (>100 items)
when credentials span multiple pages.

Screencasts:
[Screencast from 2025-09-22 00-18-49.webm](https://github.com/user-attachments/assets/87bff55b-f187-4c8d-b5cc-61652f8a5800)

[Screencast from 2025-09-22 00-17-42.webm](https://github.com/user-attachments/assets/919ad873-640b-4d67-ae14-681f4409ac81)

## Summary by Sourcery

Enable scalable credential selection by adding a multi-select typeahead component with server-side search, debounced calls, pagination, and selection caching, and integrate it into the Add Source modal.

New Features:
- Introduce TypeaheadCheckboxesWithSearch component with server-side search, debounced requests, pagination support, and selection persistence
- Add useGetCredentialsWithSearchApi hook for fetching credentials via server-side search

Enhancements:
- Replace local credential filtering with server-side search_by_name parameter and maintain a credentials cache across pages
- Integrate new search-enabled typeahead into Add Source modal and load initial selected credentials in edit mode

Tests:
- Add unit tests and snapshots for useGetCredentialsWithSearchApi hook
- Add comprehensive tests for TypeaheadCheckboxesWithSearch covering search, pagination, selection, and placeholder behaviors
- Update Add Source modal tests to reflect the new initialSelectedCredentials prop

infinitewarp reviewed (changes requested):
As I mentioned in Slack a couple days ago, but I forgot to say here, I am happy with what I've seen in the demo and your explanations for this new behavior. 👍 

I am not an expert in the surrounding code, and what I've read here seems reasonable, but the one issue I noticed (and mentioned in an inline comment) is a possible lack of string localization. So, I'm marking this with "request changes" for now, but please let me know if I just misunderstand how the code works, and I'll flip back to "approve".

nicolearagao commented on src/components/typeAheadCheckboxes/typeaheadCheckboxes.tsx:
```diff
@@ -15,83 +22,254 @@ import {
+  const [activePlaceholder, setActivePlaceholder] = useState(defaultPlaceholder);
+  const [isLoading, setIsLoading] = useState(false);
+  const [hasMorePages, setHasMorePages] = useState(false);
+  const [availableCredentials, setAvailableCredentials] = useState<CredentialOption[]>([]);
```
sure, I chose displayedCredentials as I believe it would be the most accurate as it reflects the component's responsibility of managing what the user sees.

nicolearagao commented on src/components/typeAheadCheckboxes/typeaheadCheckboxes.tsx:
```diff
@@ -15,83 +22,254 @@ import {
+
+    // Don't set allCredentials here - we need to fetch all available credentials
+    // setAllCredentials should only be set when we fetch the complete list from the API
+  }, [initialSelectedCredentials?.length]);
```
I guess my intention behind it was preventing the effect from running when parent re-renders and creates new array references, but that was misguided. I updated it to use the array instead.
The issue with using .length is that it can lead to stale state, if the parent component changes the array contents but keeps the same length (like replacing one credential with another), the effect won't re-run and the component state won't update properly. I haven't thought of that before.

nicolearagao commented on src/components/typeAheadCheckboxes/typeaheadCheckboxes.tsx:
```diff
@@ -15,83 +22,254 @@ import {
+        performLocalFilter(inputValue);
+      } else {
+        // Use server-side search when we don't have all data yet or when dropdown first opens
+        debouncedSearch(inputValue);
       }
     }
+  }, [inputValue, isOpen, debouncedSearch, hasMorePages, allCredentials.length, performLocalFilter]);
```
I've implemented the fix by removing hasMorePages from the useEffect dependencies, and it's working well.
 I confirmed this fix by :
 - Single API request when opening dropdown (was 2 before, now it's one)
 - Single API request when filtering (was 2 before, now it's one). While testing this I also increaded the debounce delay to 600 ms.

nicolearagao commented:
> One more thing: this is somewhat related to environment and might be hard to reproduce, but with data I have, item from the last page of results might jump in ahead of other items after being selected. Hopefully this video makes it clear:
> 
>  vokoscreenNG-2025-09-29_17-12-50.mp4 
> Look at relative order of items. Once I select the last item, it suddenly becomes first.
> 
> What is hopefully visible in later part of video, no matter if you order by names or ids, that item should be last. So I don't really know why it jumps ahead of others. It looks like some calculation error somewhere, or perhaps in one place we use `results` list index instead of object id.

@mirekdlugosz I could reproduce locally and I believe I fixed the issue. Changed the code to remember the exact order you select items, instead of trying to group them by which page they came from. Items will appear in the dropdown in the order you select them, that should prevent the jumping around.

mirekdlugosz reviewed (approved):
Thanks for addressing all the comments. Current version looks and works well :+1:

infinitewarp reviewed (approved):
Looks good to me. Thanks for handling the requested fixes!
